### PR TITLE
FsaGraphBuilder with HMM/CTC/RNA topology + optional minDuration

### PR DESCRIPTION
**Pull request by ZhouW321**

### CODE REVIEW

**Comment by curufinwe** [comment]
* remove configurable index for blank from state-graph-builder
* add new special lemma to lexicon for blank
* Add `Mm::EmissionIndex getBlankIndex()` to `ClassicAcousticModel`

**Comment by ZhouW321** [comment]
> * remove configurable index for blank from state-graph-builder
> * add new special lemma to lexicon for blank
> * Add `Mm::EmissionIndex getBlankIndex()` to `ClassicAcousticModel`

As a record: this requests to introduce blank label as a special lemma/phon from the lexicon and its classID will be determined via state-tying. 
correction: it should be blank `allophoneStateIndex` for the `ClassicAcousticModel`, not `emissionIndex`
(all done)

**Comment by SimBe195** [comment]
I built a small test training setup to check the final generated graph and it seems to be wrong at the moment. I suspect that the conditions at "skip loop and useless arcs for blank" are incorrect. Blanks are inserted only between words and minimum duration also is only applied to blank; except for the one at sentence end. Here is an image of the generated graph for the phrase "This week" = "dh ih s  w ee k":

graphviz

If you need the test setup for debugging I can give it to you.

**Comment by ZhouW321** [comment]
I need to see your config first,  yes, let's have a look at your setup offline

**Comment by ZhouW321** [comment]
ok, there were indeed some bugs. All should be solved now, together with a bit more logs/plots
here are the outputs of your example setup:
CTC
image

CTC + min-duration 2
image

**Review by SimBe195 — APPROVED** [review]

File: src/Speech/AllophoneStateGraphBuilder.cc [hunk 1]
```diff
@@ -354,3 +350,211 @@ Fsa::ConstAutomatonRef AllophoneStateGraphBuilder::createAlignmentGraph(const Al
...
+    return automaton;
+}
+
+// -------- CTC Topology --------
+CTCTopologyGraphBuilder::CTCTopologyGraphBuilder(const Core::Configuration& config,
+                                                 Core::Ref<const Bliss::Lexicon> lexicon,
+                                                 Core::Ref<const Am::AcousticModel> acousticModel,
+                                                 bool flatModelAcceptor) :
+        Precursor(config, lexicon, acousticModel, flatModelAcceptor),
+        labelLoop_(true),
```
[the remark this answers is not in the document]
Comment (ZhouW321) [reply src/Speech/AllophoneStateGraphBuilder.cc:424]: avoid redundant code should be preferred. From a conceptual level, the fundamental difference between CTC and RNA topology is loop-or-not.  This can be achieved by either this flag labelLoop_  or   introducing a virtual funtion, e.g. addLoopTransition, for that if (labelLoop_) part.
Probably the latter is better as we can further unfiy the code also for HMMTopology.

**Comment by ZhouW321** [comment]
check this out

**Review by SimBe195 — APPROVED** [review]

**Review by curufinwe — APPROVED** [review]
